### PR TITLE
Add code for cloud shell managed identity

### LINKED ISSUE #3832: [Feature Request] Add Cloud Shell managed identity source to MSI Integration
Add Cloud Shell managed identity source.

- [x] Add code for Cloud Shell source
- [x] Add unit tests 
- [x] Add error handling 
- [ ] Manual testing

For more details see: #3754

### PR DESCRIPTION
**Pull request by neha-bhargava**
Fixes #3832 

**Changes proposed in this request**
Add code for cloud shell managed identity

**Testing**
Unit tests

**Performance impact**

**Documentation**
- [ ] All relevant documentation is updated.

### CODE REVIEW

File: src/client/Microsoft.Identity.Client/ManagedIdentity/AppServiceManagedIdentitySource.cs [hunk 1]
```diff
@@ -19,8 +19,6 @@ internal class AppServiceManagedIdentitySource : ManagedIdentitySource
         private const string AppServiceMsiApiVersion = "2019-08-01";
         private const string SecretHeaderName = "X-IDENTITY-HEADER";
 
-        private const string MsiEndpointInvalidUriError = "[Managed Identity] The environment variable IDENTITY_ENDPOINT contains an invalid Uri {0} in app service managed identity source.";
```
Comment (trwalke) [inline src/client/Microsoft.Identity.Client/ManagedIdentity/AppServiceManagedIdentitySource.cs:22]: Should this be in MsalErrorMessages?

Comment (neha-bhargava) [reply src/client/Microsoft.Identity.Client/ManagedIdentity/AppServiceManagedIdentitySource.cs:22]: It is moved to MsalErrorMessage and removed from here.

File: src/client/Microsoft.Identity.Client/MsalErrorMessage.cs [hunk 2]
```diff
@@ -446,10 +446,10 @@ public static string InvalidTokenProviderResponseValue(string invalidValueName)
...
+        public const string ManagedIdentityEndpointInvalidUriError = "[Managed Identity] The environment variable {0} contains an invalid Uri {1} in {2} managed identity source.";
         public const string ManagedIdentityNoChallengeError = "[Managed Identity] Did not receive expected WWW-Authenticate header in the response from Azure Arc Managed Identity Endpoint.";
         public const string ManagedIdentityInvalidChallange = "[Managed Identity] The WWW-Authenticate header in the response from Azure Arc Managed Identity Endpoint did not match the expected format.";
-        public const string ManagedIdentityUserAssignedNotSupported = "[Managed Identity] User assigned identity is not supported by the Azure Arc Managed Identity Endpoint. To authenticate with the system assigned identity omit the client id to .WithManagedIentity().";
+        public const string ManagedIdentityUserAssignedNotSupported = "[Managed Identity] User assigned identity is not supported by the Azure Arc and Cloud Shell Managed Identity Endpoints. To authenticate with the system assigned identity omit the client id to .WithManagedIentity().";
```
Comment (gladjohn) [inline src/client/Microsoft.Identity.Client/MsalErrorMessage.cs:452]:
```suggestion
        public const string ManagedIdentityUserAssignedNotSupported = "[Managed Identity] User assigned identity is not supported by the Azure Arc and Cloud Shell Managed Identity Endpoints. To authenticate with the system assigned identity use .WithManagedIdentity().";

```

File: src/client/Microsoft.Identity.Client/ManagedIdentity/CloudShellManagedIdentitySource.cs [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+            }
+
+            requestContext.Logger.Verbose("[Managed Identity] Creating cloud shell managed identity. Endpoint URI: " + msiEndpoint);
+            return new CloudShellManagedIdentitySource(endpointUri, requestContext);
+        }
+
+        private CloudShellManagedIdentitySource(Uri endpoint, RequestContext requestContext) : base(requestContext)
+        {
+            _endpoint = endpoint;
+            if (!string.IsNullOrEmpty(requestContext.ServiceBundle.Config.ManagedIdentityUserAssignedClientId) || null != requestContext.ServiceBundle.Config.ManagedIdentityUserAssignedResourceId)
```
Comment (pmaytak) [inline src/client/Microsoft.Identity.Client/ManagedIdentity/CloudShellManagedIdentitySource.cs:49]: Thoughts on refactoring these two properties? Combine them into one, something like `UserAssignedManagedIdentityId` and when it's added to the request, if it's a guild add it as client ID, or resource ID. We don't really need to differentiate anywhere else in the code.

**Review by pmaytak — APPROVED** [review]

**Review by trwalke — APPROVED** [review]

**Review by gladjohn — APPROVED** [review]

File: src/client/Microsoft.Identity.Client/MsalErrorMessage.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (neha-bhargava) [reply src/client/Microsoft.Identity.Client/MsalErrorMessage.cs]: Do you mean to create constants for the names of the sources?

[the remark this answers is not in the document]
Comment (neha-bhargava) [reply src/client/Microsoft.Identity.Client/MsalErrorMessage.cs]: Resource would be better I think.

File: src/client/Microsoft.Identity.Client/ManagedIdentity/CloudShellManagedIdentitySource.cs (hunk 3 above)
Comment (neha-bhargava) [reply src/client/Microsoft.Identity.Client/ManagedIdentity/CloudShellManagedIdentitySource.cs]: Updated.

**Pull request merged**